flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java

```diff
@@ -260,36 +259,44 @@ public void releaseResource(YarnWorkerNode workerNode) {
 	//  Internal
 	// ------------------------------------------------------------------------
 
-	private void onContainersOfResourceAllocated(Resource resource, List<Container> containers) {
-		final List<TaskExecutorProcessSpec> pendingTaskExecutorProcessSpecs =
-			taskExecutorProcessSpecContainerResourceAdapter.getTaskExecutorProcessSpec(resource, matchingStrategy).stream()
-				.flatMap(spec -> Collections.nCopies(getNumRequestedNotAllocatedWorkersFor(spec), spec).stream())
-				.collect(Collectors.toList());
+	private void onContainersOfPriorityAllocated(Priority priority, List<Container> containers) {
+		final Optional<TaskExecutorProcessSpecContainerResourcePriorityAdapter.TaskExecutorProcessSpecAndResource> taskExecutorProcessSpecAndResourceOpt =
+			taskExecutorProcessSpecContainerResourcePriorityAdapter.getTaskExecutorProcessSpecAndResource(priority);
+
+		if (!taskExecutorProcessSpecAndResourceOpt.isPresent()) {
+			log.warn("Receive {} containers with unrecognized priority {}. This should not happen.",
+				containers.size(), priority.getPriority());
+			for (Container container : containers) {
+				returnExcessContainer(container);
+			}
+			return;
+		}
```

Comment:
Are we not failing hard here because this can happen in case of a JobManager failover?

Reply:
Actually I don't see any possible reasons that `taskExecutorProcessSpecAndResourceOpt` does not exist. This is just a safety net. Maybe we can just replace this with `checkState`.

Reply:
Then I would vote for `checkState` to impose stricter conditions and to fail fast in case our assumptions are broken.